Source/WebKit/Shared/cf/CoreIPCSecTrust.mm

```diff
@@ -0,0 +1,713 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if HAVE(WK_SECURE_CODING_SECTRUST) && USE(CF)
+
+#import "config.h"
+#import "CoreIPCSecTrust.h"
+
+#import "Logging.h"
+#import <wtf/Compiler.h>
+#import <wtf/cocoa/TypeCastsCocoa.h>
+
+namespace WebKit {
+
+
+static bool arrayElementsTheSameType(NSArray *array)
+{
+    if (!array.count)
+        return true;
+    Class firstElement = [array.firstObject class];
+    for (id element in array) {
+        if (![element isKindOfClass:firstElement])
+            return false;
+    }
+    return true;
+}
+
+CoreIPCSecTrust::PolicyOptionValueShape CoreIPCSecTrust::detectPolicyOptionShape(id option)
+{
+    if ([option isKindOfClass:NSNumber.class]) {
+        NSNumber *candidateBool = option;
+        if ([candidateBool isEqualToNumber:@YES] || [candidateBool isEqualToNumber:@NO])
+            return PolicyOptionValueShape::Bool;
+    } else if ([option isKindOfClass:NSString.class])
+        return PolicyOptionValueShape::String;
+    else if ([option isKindOfClass:NSArray.class]) {
+        NSArray *a = option;
+        id element = a.firstObject;
+        if (!element) {
+            RELEASE_LOG_ERROR(IPC, "CoreIPCSecTrust::detectPolicyOptionShape element was nil");
+            ASSERT_NOT_REACHED();
+            return PolicyOptionValueShape::Invalid;
+        }
+        if ([element isKindOfClass:NSNumber.class])
+            return PolicyOptionValueShape::ArrayOfNumbers;
+        if ([element isKindOfClass:NSString.class])
+            return PolicyOptionValueShape::ArrayOfStrings;
+        if ([element isKindOfClass:NSData.class])
+            return PolicyOptionValueShape::ArrayOfData;
+        if ([element isKindOfClass:NSArray.class])
+            return PolicyOptionValueShape::ArrayOfArrayContainingDateOrNumber;
+    } else if ([option isKindOfClass:NSDictionary.class])
+        return PolicyOptionValueShape::DictionaryValueIsNumber;
+
+    RELEASE_LOG_ERROR(IPC, "CoreIPCSecTrust::detectPolicyOptionShape was unable to detect option's shape");
+    ASSERT_NOT_REACHED();
+    return PolicyOptionValueShape::Invalid;
+}
+
+static bool updatePolicyVector(NSDictionary *policyOption, CoreIPCSecTrustData::PolicyOption& policyVector)
```

Comment:
We could make `updatePolicyVector` return an error `String`, then on the caller side you could check the return value, and if it's not null, print it and early return.